### PR TITLE
Marisa Morris - Carets

### PR DESCRIPTION
**Pull request by marisamorris**
# Hotel
Congratulations! You're submitting your assignment!

## Comprehension Questions
| Question | Answer    |
| :------------- | :------------- |
| Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? |    Whether or not to initialize the rooms automatically or use a helper method due to not every hotel having twenty rooms. I chose to use a method to create the rooms because I figured if this project is ever built on, it should be able to be changed easily.   |
| Describe a concept that you gained more clarity on as you worked on this assignment. |  Using inheritance as well as how to use the built in Date object.  |
| Describe a _nominal test_ that you wrote for this assignment. |  Being able to instantiate each class.  |
| Describe an _edge case test_ that you wrote for this assignment. | An edge case would be if the check in date is in the past or the check out date is before the check in date for the create reservation method  |
| How do you feel you did in writing pseudocode first, then writing the tests and then the code? | Writing pseudo code then the tests were much more helpful in that it helped me figure out what i ultimately wanted each class and method to do in addition to where i needed to refactor my Hotel class.|

### CODE REVIEW

File: lib/block_reservation.rb [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+module HotelManagment
+  class BlockReservation < Reservation
+
+    def initialize(first_name, last_name, check_in_date, check_out_date, room_number)
+      super
+      @cost = cost
+    end
+
+    def block_rate
```
Comment (CheezItMan) [inline lib/block_reservation.rb:12]: Shouldn't you simply override the `rate` method?

Why create the same method by a different name in a subclass?

File: lib/reservation.rb [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+      @last_name = last_name
+      @check_in_date = check_in_date
+      @check_out_date = check_out_date
+      @room_number = room_number
+      @cost = cost
+    end
+
+    def rate
+      nights = @check_out_date - @check_in_date
+      @cost = nights.to_i * Room::RATE
```
Comment (CheezItMan) [inline lib/reservation.rb:21]: You set the cost to a different value in `initialize`.  Why?

File: lib/room.rb [hunk 3]
```diff
@@ -0,0 +1,15 @@
+#room
+require 'awesome_print'
+
+module HotelManagment
+  class Room
+    attr_reader :room_number
+
+    RATE = 200
```
Comment (CheezItMan) [inline lib/room.rb:8]: Good use of a constant!

File: lib/hotel.rb [hunk 4]
```diff
@@ -0,0 +1,119 @@
+# hotel
+require_relative 'room.rb'
+require 'date'
+
+module HotelManagment
+  class Hotel
+    attr_reader :rooms, :reservations, :found_reservations, :unreserved_rooms, :blocks
+    def initialize
+      @rooms = []
```
Comment (CheezItMan) [inline lib/hotel.rb:9]: If you're setting `@rooms` to `[]`  in initialize, why do you have an `add_20_rooms` method.  I would suggest instead making `add_20_rooms` private and calling it from `initialize`.  I would also suggest renaming it to something less specific as to the implementation (number of rooms).  That way you could extend or subclass `Hotel` and override the method to change the number of rooms.

**Comment by CheezItMan** [comment]
# Hotel
## What We're Looking For

Feature | Feedback
---     | ---
**Baseline** |
Used git regularly	| Decent number of commits, good messages |
Answer comprehension questions	| Check |
**Design** |
Each class is responsible for a single piece of the program | Most of the program is in the `Hotel` class.  You need to break out the functionality more.    |
Classes are loosely coupled | Everything in this program revolves around the `Hotel` class, The other classes don't have much functionality.   |
**Wave 1** |
List rooms |  |
Reserve a room for a given date range |  Check via the `rooms_not_reserved` method |
List reservations for a given date | Check via `reservations_by_date`. |
Calculate reservation price | Your reservation class has the method, see my notes about it. |
Invalid date range produces an error | Check |
**Wave 2** |
View available rooms for a given date range | Check  |
Reserving a room that is not available produces an error | Nope:  `create_reservation` doesn't check to ensure the room hasn't already been reserved! |
**Wave 3** |
Create a block of rooms | There' not check here for dates that are invalid.   |
Check if a block has rooms | There's a way to reserve a room from a block, but no way to see if a block has rooms.   |
Reserve a room from a block | Via the `reserve_room_in_block` method.  There are no tests for the dates.  See my notes on these classes.   |
Test coverage |  99% ! | 
**Additional Feedback** | You are putting almost all your business logic into `Hotel` and not the other classes.  Those other classes seem mostly to be used to store data.  This makes it a very difficult design to change, modify or reuse components.  For things like `Reservation` you test the data types of attributes and instantiating, but no tests for edge cases.  This is problematic.  I'd like you to go over your design with your tutor and your tests as well and talk about what you could do differently and how to write tests for a class.  This was a tough project and it looks like you learned a lot.  Also look at your hotel class and look at the loops in some of the methods, they look like they repeat a lot, you could do some DRYing up here.    |

**Review by CheezItMan — COMMENTED** [review]
Notes on your code.

File: lib/hotel.rb [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+        room = HotelManagment::Room.new(room_number)
+        @rooms << room
+        room_number += 1
+      end
+    end
+
+
+    # creates a reservation
+    def create_reservation(first_name, last_name, check_in_date, check_out_date, room_number)
+
```
Comment (CheezItMan) [inline lib/hotel.rb:30]: The `Reservation` class should be checking for it being created with invalid dates.

File: lib/hotel.rb [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+      reservation = Reservation.new(first_name, last_name, check_in_date, check_out_date, room_number)
+
+      @reservations << reservation
+    end
+
+
+
+    def reservations_by_date(date)
+      @reservations.each { |reservation|
```
Comment (CheezItMan) [inline lib/hotel.rb:39]: Good use of enumerables in this project.
